flocker/node/test/test_deploy.py

```diff
@@ -2291,13 +2336,371 @@ def test_dataset_id_generated(self):
+            network=make_memory_network()
+        )
+
+        calculating = api.calculate_necessary_state_changes(
+            desired_state=desired,
+            current_cluster_state=current,
+            hostname=u"node1.example.com",
+        )
+
+        dataset = MANIFESTATION_WITH_SIZE.dataset
+
+        changes = self.successResultOf(calculating)
+        # expected is: resize, push, handoff
+        expected = Sequentially(changes=[
+            InParallel(
+                changes=[ResizeDataset(dataset=dataset)],
+            ),
+            InParallel(
+                changes=[PushDataset(
+                    dataset=dataset,
+                    hostname=u'node2.example.com')]
+            ),
+            InParallel(
+                changes=[HandoffDataset(
+                    dataset=dataset,
+                    hostname=u'node2.example.com')]
+            )])
+        self.assertEqual(expected, changes)
+
+    def test_dataset_max_size_preserved_after_move(self):
+        """
+        ``Deployer.calculate_necessary_state_changes`` specifies that a
+        dataset will be resized if it is to be relocated to a different
+        node but specifies a maximum_size that differs to the existing
+        size. The dataset on the new target node will be resized after it
+        has been received.
+        """
+        volume_service = create_volume_service(self)
+        self.successResultOf(volume_service.create(
+            volume_service.get(
+                _to_volume_name(DATASET.dataset_id))
+            )
+        )
+        docker = FakeDockerClient(units={})
+
+        current_nodes = [
+            Node(
+                hostname=u"node1.example.com",
+                other_manifestations=frozenset([MANIFESTATION]),
+            ),
+            Node(
+                hostname=u"node2.example.com",
+            )
+        ]
+
+        desired_nodes = [
+            Node(
+                hostname=u"node1.example.com",
+            ),
+            Node(
+                hostname=u"node2.example.com",
+                other_manifestations=frozenset({MANIFESTATION_WITH_SIZE}),
+            ),
+        ]
+
+        current = Deployment(nodes=frozenset(current_nodes))
+        desired = Deployment(nodes=frozenset(desired_nodes))
+
+        api = Deployer(
+            volume_service, docker_client=docker,
+            network=make_memory_network()
+        )
+
+        calculating = api.calculate_necessary_state_changes(
+            desired_state=desired,
+            current_cluster_state=current,
+            hostname="node2.example.com",
+        )
+
+        changes = self.successResultOf(calculating)
+        dataset = MANIFESTATION_WITH_SIZE.dataset
+
+        expected = Sequentially(changes=[
+            InParallel(changes=[WaitForDataset(dataset=dataset)]),
+            InParallel(changes=[ResizeDataset(dataset=dataset)]),
+        ])
+        self.assertEqual(expected, changes)
+
+
```

Comment:
Not exercised by this `TestCase` is the case where `primary` changes value.  Though `primary` can't ever be anything but `True` now so that's probably okay.  And we don't have an `IStateChange` to handle that either.  And maybe the ZFS backend doesn't do anything for that change anyway (or maybe it switches between being mounted read-write and read-only?).

So maybe there's nothing to talk about here.  I guess I'm still not entirely clear how `Manifestation.primary` is going to interact with the rest of the system.  I guess it's derived from some configuration we store somewhere... but we don't have that yet, nor specs for any APIs that might manipulate it.  It's probably related to two-phase push, though, somehow - which we do have ...  Hm.

Can state discovery produce `primary=False` `Manifestations`?  (It can't but maybe it should be able to?  The ZFS dataset left after a volume "moves" to another node is probably not a `primary=True` `Manifestation` ...)